Shared/IISFunctions/Get-IISModules.ps1

```diff
@@ -0,0 +1,159 @@
+                                        Write-Verbose "Signer information found. Subject: $($signature.SignerCertificate.Subject)"
+                                        $mDetails.Signer = $signature.SignerCertificate.Subject.ToString()
+                                        $mDetails.IsMicrosoftSigned = $signature.SignerCertificate.Subject -cmatch "O=Microsoft Corporation, L=Redmond, S=Washington"
+                                    }
+
+                                    $mInfoObject.SignatureDetails = $mDetails
+                                }
+
+                                $iisModulesList.Add($mInfoObject)
+                            } catch {
+                                Write-Verbose "Unable to validate file signing information"
+                                if ($null -ne $CatchActionFunction) {
+                                    & $CatchActionFunction
+                                }
+                            }
+                        }
+                    } else {
+                        Write-Verbose "No modules are loaded by IIS"
+                    }
+                } catch {
+                    Write-Verbose "Failed to process global module information. $_"
+                    if ($null -ne $CatchActionFunction) {
+                        & $CatchActionFunction
+                    }
+                }
+            }
+            end {
+                return $iisModulesList
+            }
+        }
+    }
+    process {
+        $modules = GetIISModulesLoaded -Xml $ApplicationHostConfig
+
+        # Validate that all modules are signed by Microsoft Corp.
+        if ($modules.SignatureDetails.IsMicrosoftSigned.Contains($false)) {
+            Write-Verbose "At least one module which is loaded by IIS is not digitally signed by 'Microsoft Corporation'"
+            $allModulesSignedByMSFT = $false
+        } else {
+            Write-Verbose "All modules which are loaded by IIS are signed by 'Microsoft Corporation'"
+            $allModulesSignedByMSFT = $true
+        }
+
+        # Validate if all signatures are valid (regardless of whether signed by Microsoft Corp. or not)
+        $allSignaturesValid = $true
+        foreach ($m in $modules) {
+            Write-Verbose "Module: $($m.Name) Signature Status: $($m.SignatureDetails.SignatureStatus)"
+            if (($m.Signed -eq $true) -and
+                ($m.SignatureDetails.SignatureStatus -ne 0)) {
+                $allSignaturesValid = $false
+            }
+        }
+
+        # Validate if all modules that are loaded are digitally signed
+        if ($modules.Signed.Contains($false)) {
+            Write-Verbose "At least one module which is loaded by IIS is not digitally signed"
+            $allModulesAreSigned = $false
+        } else {
+            Write-Verbose "All modules which are loaded by IIS are digitally signed"
+            $allModulesAreSigned = $true
+        }
```

Comment:
This could be all one loop vs 3 different ones

Reply:
Simplified the code: 
<img width="650" alt="image" src="https://user-images.githubusercontent.com/40993616/182917994-7b7fe259-b031-4776-8d9a-bdf21d8d1a42.png">